node and wallet: expose spv and prune settings

pinheadmz opened the pull request:
This is cleaner refactor of #532, which I'm separating into two parts. Part one is this PR, which exposes the node's chain's `spv` and `prune` settings in a few places to make it more accessible to wallets. (Part two will use this data to prevent a rescan beyond available blocks in pruned mode, and will switch to chain reset/replay for spv nodes).

It will also provide better support for features like https://github.com/bcoin-org/bcoin/pull/621

OrfeasLitos commented on lib/wallet/nodeclient.js:
```diff
@@ -211,6 +211,20 @@ class NodeClient extends AsyncEmitter {
+
+  /**
+  * Get relevant node info for wallet
+  * @returns {Promise}
```
Why is getInfo() async? Seems redundant, given that it returns a resolved promise and it's `await`ed for in all the places where it's called.

pinheadmz replied:
Thanks Orfeas, just cleaned this up.

OrfeasLitos commented on lib/wallet/node.js:
```diff
@@ -108,6 +108,12 @@ class WalletNode extends Node {
     await this.http.open();
     await this.handleOpen();
 
+    const clientInfo = await this.client.getInfo();
```
`await` unneeded

pinheadmz replied:
Can't remove it here, if wallet is running as a standalone service `getInfo()` is an actual HTTP API call. If you remove it here, the test will fail.

OrfeasLitos commented on lib/wallet/nodeclient.js:
```diff
@@ -211,6 +211,20 @@ class NodeClient extends AsyncEmitter {
+
+  /**
+  * Get relevant node info for wallet
+  * @returns {Promise}
```
doesn't return Promise anymore

pinheadmz replied:
👍

OrfeasLitos commented on lib/wallet/nullclient.js:
```diff
@@ -162,6 +162,20 @@ class NullClient extends EventEmitter {
+
+  /**
+  * Get relevant node info for wallet
+  * @returns {Promise}
```
doesn't return Promise anymore

pinheadmz replied:
:+1:

OrfeasLitos commented on test/wallet-test.js:
```diff
@@ -245,6 +245,12 @@ describe('Wallet', function() {
   });
 
+  it('should get dummy chain info (from NullClient)', async () => {
+    const getInfo = await wdb.client.getInfo();
```
`await` unneeded

pinheadmz replied:
:+1: In this case no you're right thanks

pinheadmz commented:
@OrfeasLitos cleaned up `await` where applicable, squashed an updated. Thanks again for the review!

tynes commented on lib/wallet/client.js:
```diff
@@ -77,6 +77,10 @@ class WalletClient extends NodeClient {
 
     return super.rescan(start);
   }
+
```
Interesting how the `async` is used in some cases and not in other cases. I think you should make this method `async` because explicit is better than implicit.

https://github.com/bcoin-org/bclient/blob/master/lib/node.js#L61

tynes commented on lib/wallet/walletdb.js:
```diff
@@ -196,6 +196,12 @@ class WalletDB extends EventEmitter {
       this.state.startHeight);
 
+    const clientInfo = await this.client.getInfo();
+    this.chainInfo = {
```
There is no null value for `chainInfo` in the constructor

pinheadmz replied:
ah thanks 👍

pinheadmz replied:
Should it be `null` or an expected default value like
```
{
  spv: false
  prune: false
}
```

tynes replied:
I think it should at least be an object because we do not want to accidentally cause a problem where a property on `null` is referenced

I think the expected default value makes sense in this case

pinheadmz replied:
went with `{}` for null value in case we want to expand, it's one less place to have to update. and it'll return `undefined` instead of throwing for requested properties

tynes replied:
👍 good point

tynes commented:
I noticed that the same chunk of code is repeated in 3 places:

```
    const clientInfo = this.client.getInfo();
    this.chainInfo = {
      spv: clientInfo.chain.spv,
      prune: clientInfo.chain.prune
    };
```

It would be most ideal if this could only be done in one place. If we want to update this interface in the future, that would be annoying to have to change in multiple places. Over the longterm, I could see other useful information being returned from a `wallet.getInfo` call

pinheadmz commented:
@tynes the three places thing is because the wallet connects to the node in three ways: as a plugin, as a standalone server, and (in tests) to a `nullclient`. Anything we do that touches wallet/node communication has to be done thrice.

This is also why _sometimes_ `getInfo()` is async. As @OrfeasLitos pointed out, I had `async` in a few places where there were no promises at all. After cleaning up on a recent commit, now its only `async` when an actual http API call is made (wallet as remote server). Otherwise, the data can be pulled directly from `this.node.chain....` etc.

The only place where its ambiguous now is in `walletDB`. There is an `async` there, but the value of `client` could be any of the three object listed above, and in only one of those, `getInfo()` is asyncronous.

pinheadmz commented:
Thanks to advice from @tynes, this PR is much simpler. `prune` and `spv` data from `node.chain` are just stored in walletDB now, they don't need to be stored anywhere else.

braydonf commented on lib/wallet/walletdb.js:
```diff
@@ -196,6 +197,12 @@ class WalletDB extends EventEmitter {
       this.state.height,
       this.state.startHeight);
 
+    const clientInfo = await this.client.getInfo();
```
This isn't used for anything in walletdb at this point, that looks to be a part of https://github.com/bcoin-org/bcoin/pull/751 rather.

braydonf commented on lib/node/http.js:
```diff
@@ -125,7 +125,9 @@ class HTTP extends Server {
           tip: this.chain.tip.rhash(),
-          progress: this.chain.getProgress()
+          progress: this.chain.getProgress(),
+          spv: this.chain.options.spv,
```
SPV is more a property of the node as it affects other features of the node too and not just the chain, and similar to pruned. What about here?
```js
{
  "version": "1.0.2",
  "network": "regtest",
  "chain": {
    "height": 200,
    "tip": "03c1d2ebd85058180da927dc66ad3fb862dd9e188f3de6d0b53161d538a382cc",
    "progress": 0.9975950784699014
  },
  // (...)
  "prune": false,
  "spv": false
}
```

Via this in lib/node/http.js:
```js
  // (...)
  },
  memory: this.logger.memoryUsage(),
  prune: this.node.prune,
  spv: this.node.spv,
});
```
The property should already be available for `node.spv`, and `node.pruned` would need to be, it's used multiple places so would be a good move to make.
```

braydonf replied:
Would it be worth including the pruned depth? It's a constant network property now, but it seems like something that would be configurable.

pinheadmz replied:
This seems like a good idea to have `prune` set at the node level like `spv` and reported at the root of the `getInfo` object, will update. Prune HEIGHT is available in `rpc getblockchaininfo` the prune depth is set by `keepBlocks` in `networks.js` for now. It's set to 288 blocks for mainnet and that is the standard minimum proposed by [BIP159](https://github.com/bitcoin/bips/blob/master/bip-0159.mediawiki) for `NODE_NETWORK_LIMITED` peers, which we don't actually implement yet, but will soon :-)

braydonf commented on lib/wallet/nodeclient.js:
```diff
@@ -211,6 +211,20 @@ class NodeClient extends AsyncEmitter {
+  */
+
+  getInfo() {
+    return {
```
The result would be different here between `client.js` and `nodeclient.js`. For example, `client.js` would give back _all_ of the info, and `nodeclient.js` gives back only this portion. The code in `http.js` for `get('/'` could be moved to `node.getInfo` and both could call it, and thus the results would be the same. Alternatively, the client could have `isPruned` and `isSPV` methods, and likewise the results would be the same for both `nodeclient.js` and `client.js`.

braydonf replied:
The result would be different here between `client.js` and `nodeclient.js`. For example, `client.js` would give back _all_ of the info, and `nodeclient.js` gives back only this portion. The code in `http.js` for `get('/'` could be moved to `node.getInfo` and both could call it, and thus the results would be the same. Alternatively, the client could have `isPruned` and `isSPV` methods, and likewise the results would be the same for both `nodeclient.js` and `client.js`.

pinheadmz commented:
@braydonf addressed your last two comments with an additional commit, take a look and if you think its the right approach I'll squash.

Updates:
1) `getInfo` returns `prune` and `spv` at the root level of the JSON object
2) `prune` setting is available at `node.prune` similar to `node.spv`
3) `getInfo()` in `client.js` now matches returned value from `nodeclient.js`  which is JUST an object with two values i.e. `{spv: false, prune: true}`

braydonf commented on CHANGELOG.md:
```diff
@@ -43,6 +43,13 @@ take multiple hours (e.g. 8 hours) depending on hardware and the
+
+#### HTTP
+
+- `GET /` (get info) Return object `chain` now includes boolean properties `spv`
```
nit: outdated now

braydonf commented on lib/wallet/walletdb.js:
```diff
@@ -196,6 +197,12 @@ class WalletDB extends EventEmitter {
       this.state.height,
       this.state.startHeight);
 
+    const clientInfo = await this.client.getInfo();
```
nit: this is node info and not chain (as the last change)

braydonf replied:
Also, this can probably be a one-liner now.

pinheadmz commented:
@braydonf thanks again, cleaned this up.

braydonf commented on lib/node/fullnode.js:
```diff
@@ -43,6 +43,9 @@ class FullNode extends Node {
     this.spv = false;
 
+    // Prune flag
+    this.prune = this.config.bool('prune');
```
nit: `this.config.bool('prune')` is used three places for `FullNode` for `Chain`, `TXIndexer`, `AddrIndexer` and could use `this.prune` instead, and the value could default to `false` with `this.config.bool('prune', false)`, it's current value is `null`. The abstract node at `lib/node/node.js` should probably have `this.prune = false`, similar to the `this.spv = false`.

pinheadmz replied:
Yes. Great idea, adding.

pinheadmz commented:
@braydonf addressed the nit around `this.prune` in `fullnode` and default value for abstract `node` -- also added lines to the tests to make sure default values came through, so spv and prune tests both check for spv and prune. Wondering what you think about those test file names now: `spv-node-test` and `prune-node-test`. Is that OK or should we be more specific ie `spv-node-conf-test` or something? Since we're not really checking anything else about the nodes yet.

braydonf commented:
The tests could be included in `node-test.js` and `wallet-test.js`?

pinheadmz commented:
Hmmmm, ok I think then they are ok as is for now :-)

pinheadmz commented:
Ok I just realized this branch will crash if a user tries to run `bwallet` without a node:

```
[error] (node) Network error: ws://localhost:8332/socket.io/?transport=websocket: connect ECONNREFUSED 127.0.0.1:8332
Error: connect ECONNREFUSED 127.0.0.1:8332
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1097:14)
```

This is something we kinda discussed in https://github.com/bcoin-org/bcoin/pull/601

One solution is to initialize `walletDB` with actual values instead of:
```
    this.nodeInfo = {};
```
...so that could be `{spv: false, prune: false}` and then the API call in `open()` can be wrapped in a try/catch.

So I know we aren't ready for the full-on offline-wallet configuration, but I don't want to merge something that breaks existing functionality even if its bad functionality.

Thoughts?

nodar-chkuaselidze commented:
that can happen in `syncState`, every time separate wallet drops connection and reconnects to node it will verify the state. So it makes sense to only and always get the state of the node on connection.

(e.g. somehow bcoin pruned bcoin node got replaced with fullnode, but wallet is still running)

pinheadmz commented:
@nodar-chkuaselidze Check out the new configuration. `prune` and `spv` properties added to `WalletDB` object -- `spv` was already there but actually never got set or tested! The code used `this.options.spv` instead. SO, fixed that and added `prune`.

Also split up the tests, so there is new test module for standalone wallet, connecting to `full`, `prune` and `spv` nodes in sequence. The no-node test remains in `wallet-test.js` and the wallet-as-a-plugin tests are in two other files for `prune` and `spv`.

pinheadmz commented on test/wallet-standalone-test.js:
```diff
@@ -0,0 +1,86 @@
+    before(async () => {
+      await fullNode.open();
+      await wallet.open();
+      await new Promise(r => setTimeout(r, 100));
```
I tried waiting for a `'connect'` event here, but it fires inconsistently (sometimes BEFORE `wallet.open()` resolves). Went down the rabbit hole all the way through `vendor/faye-websocket` and couldn't figure out a better strategy.

nodar-chkuaselidze replied:
You can setup listener before `open`, but wait after `.open`.

If connect happened before, it will be resolved. If not it will resolve after.

pinheadmz replied:
hows this? https://github.com/bcoin-org/bcoin/pull/749/commits/d87276f5db83c5f7c6438fcba370b59c29866eab

pinheadmz commented:
Rebased on master, merge conflicts fixed.
